### PR TITLE
Show UI warning for invalid URL's

### PR DESCRIPTION
**Pull request by varjolintu**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
Shows a warning when editing entry if the URL field has an invalid value.

Also works with Additional URL's page, although the background color only changes after editing.

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
Valid URL:
Screen Shot 2019-11-22 at 13 56 59

Invalid URL:
Screen Shot 2019-11-22 at 13 57 09

Invalid URL in Additional URL's page:
Screen Shot 2019-11-22 at 23 17 21

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Tests added to `TestBrowser` class.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/gui/URLEdit.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include <QAction>
+#include <QLineEdit>
+#include <QPointer>
+
+// Use a separate class for the comparison function so tests can be run without
+// dependency to QLineEdit
+class ValidURL
+{
+public:
+    static bool validURL(const QString& urlField);
```
Comment (droidmonkey) [inline src/gui/URLEdit.h]: This is an odd place to put this function. I recommend putting this in the Tools namespace.

**Comment by droidmonkey** [comment]
Are we going to extend this to the multi url list or wait for 2.6.0?

File: src/gui/URLEdit.h (hunk 1 above)
Comment (varjolintu) [reply src/gui/URLEdit.h]: Thanks for the tip.

**Comment by varjolintu** [comment]
@droidmonkey I'll test it today and see if it's possible to do. Should be a small fix.

**Comment by varjolintu** [comment]
@droidmonkey It was possible. Although only the background color changes after editing. QModelIndex has its limits.

File: src/gui/URLEdit.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+void URLEdit::updateStylesheet()
+{
+    const QString stylesheetTemplate("QLineEdit { background: %1; }");
+
+    if (!Tools::validURL(text())) {
+        setStyleSheet(stylesheetTemplate.arg(ErrorColor.name()));        
+        m_correctAction->setVisible(false);
+        m_errorAction->setVisible(true);
+    } else {
+        m_correctAction->setVisible(false);
```
Comment (ba32107) [inline src/gui/URLEdit.cpp]: Shouldn't this be `setVisible(true)`? I don't see `m_correctAction` being made visible anywhere.

Comment (varjolintu) [reply src/gui/URLEdit.cpp]: Ah, actually I thought that we don't need toe correct action at all. It's originally only needed when the passwords match. I guess we don't want to always show a confirmation in the entry URL field if it's correct. Only show the invalid value. So the correct action member could be actually removed.

**Comment by droidmonkey** [comment]
You can show an icon in the QModelIndex, that is how I show the paperclip on the entry view model.

**Comment by varjolintu** [comment]
@droidmonkey Oh. Didn't know that. I'll add the icon.

File: src/gui/URLEdit.cpp (hunk 2 above)
Comment (ba32107) [reply src/gui/URLEdit.cpp]: Yep, I was thinking the same! 

Another thing I wanted to say is that I think this validation shouldn't actually prevent users from putting in a malformed URL, it should just be a warning, but it should be ignorable. (I didn't actually see any logic that forbids an incorrect URL, I just wanted to make my opinion explicit)

Comment (varjolintu) [reply src/gui/URLEdit.cpp]: You understood correctly, this is just a warning. It doesn't prevent using it. Still, of course it won't work properly.

**Comment by varjolintu** [comment]
Added the icon to the additional URL field's right side + removed the obsolete code from URLEdit class.

Thanks for the feedback everyone!

File: src/core/Tools.h [hunk 3]
```diff
@@ -41,6 +41,7 @@ namespace Tools
     bool isBase64(const QByteArray& ba);
     void sleep(int ms);
     void wait(int ms);
+    bool validURL(const QString& urlField);
```
Comment (droidmonkey) [inline src/core/Tools.h]: Can we name this: `checkUrlValid(...)`?

Comment (varjolintu) [reply src/core/Tools.h]: Renamed.

**Comment by varjolintu** [comment]
Removed the extra check, now tests with Linux are also fine. Everything's ready if there's nothing more to add.

**Pull request merged**